src/main/java/com/emc/object/s3/S3SignerV4.java

```diff
@@ -0,0 +1,392 @@
+        /*
+        CanonicalRequest =
+            HTTPRequestMethod + '\n' +
+            CanonicalURI + '\n' +
+            CanonicalQueryString + '\n' +
+            CanonicalHeaders + '\n' +
+            SignedHeaders + '\n' +
+            HexEncode(Hash(RequestPayload))
+         */
+        StringBuilder canonicalRequest = new StringBuilder();
+        canonicalRequest.append(request.getMethod()).append("\n");
+        URI uri = request.getURI();
+        String resource = RestUtil.getEncodedPath(uri);
+        canonicalRequest.append(resource).append("\n");
+        canonicalRequest.append(getCanonicalizedQueryString(parameters));
+
+        SortedMap<String, String> canonicalizedHeaders = getCanonicalizedHeaders(headers, parameters);
+        StringBuilder signedHeaders = new StringBuilder();
+        for (String name : canonicalizedHeaders.keySet()) {
+            canonicalRequest.append(name).append(":");
+            if (canonicalizedHeaders.get(name) != null) {
+                canonicalRequest.append(canonicalizedHeaders.get(name).trim());
+            }
+            canonicalRequest.append("\n");
+            if (signedHeaders.length() != 0)
+                signedHeaders.append(";");
+            signedHeaders.append(name);
+        }
+        canonicalRequest.append("\n");
+
+        signedHeaders.append("\n");
+        canonicalRequest.append(signedHeaders);
+
+        String hashedPayload = "";
+        String payload = "";
+        byte[] hash = hash256(payload);
+        hashedPayload = hexEncode(hash);
+        canonicalRequest.append(hashedPayload);
+        log.debug("CanonicalRequest: {}" + canonicalRequest);
+        return canonicalRequest.toString();
+    }
+
+    private String getCanonicalizedQueryString(Map<String, String> parameters) {
+        StringBuilder queryString = new StringBuilder();
+        if (parameters != null & parameters.size() != 0) {
+            SortedMap<String, String> sortedParameters = new TreeMap<String, String>();
+
+            for (Map.Entry<String, String> parameter : parameters.entrySet()) {
+                sortedParameters.put(parameter.getKey(), RestUtil.urlEncode(parameter.getValue()));
+            }
+            StringBuilder sortedQueryString = new StringBuilder();
+            for (Map.Entry<String, String> parameter : sortedParameters.entrySet()) {
+                if (sortedQueryString != null && sortedQueryString.length() != 0)
+                    sortedQueryString.append("&");
+                sortedQueryString.append(parameter.getKey()).append("=");
+                if (parameter.getValue() != null)
+                    sortedQueryString.append(parameter.getValue());
+            }
+            queryString.append(sortedQueryString).append("\n");
+        } else
+            queryString.append("\n");
+        return queryString.toString();
+    }
+
+    protected SortedMap<String, String> getCanonicalizedHeaders(Map<String, List<Object>> headers,
```

Comment:
According to the comment, this function will add x-emc-/x-amz- headers but real function is to exclude excludedSignedHeaders, what's the real desire? Is it possible to consolidate the function into the one in S3Signer.java?

Reply:
Removed unused comments. This method is already an overwrite method. V4 has different behavior on this with v2.